### DISCUSSION TITLE
How to handle timeouts

### DISCUSSION THREAD

**Discussion by Happy-Ferret** [python-telegram-bot/python-telegram-bot#2876] [Q&A] [answered]
Unfortunately, the Wiki doesn't provide any code samples on how to properly handle timeouts.

Using [echobot.py](https://github.com/python-telegram-bot/python-telegram-bot/blob/master/examples/echobot.py) as an example, connections will sometimes be closed on the Telegram end and a `NetworkError` exception is raised, hence there needs to be a retry.

It's unclear how a retry can be achieved, alas.

This is what I tried, thus far. To no avail:

```python
...
def error_handler(update: Update, context: CallbackContext) -> None:
    print("Error raised")
    update.message.chat.bot.get_updates()
    time.sleep(10)

dispatcher.add_error_handler(error_handler)
...
```

**Comment by clot27** [comment]
There's an wiki article for [Handling network errors](https://github.com/python-telegram-bot/python-telegram-bot/wiki/Handling-network-errors)

**Reply by Happy-Ferret** [reply]
> There's an wiki article for [Handling network errors](https://github.com/python-telegram-bot/python-telegram-bot/wiki/Handling-network-errors)

I know. I saw that one. It's, unfortunately, not very insightful. There's no code examples. It's all rather theoretical.
Shouldn't the examples deal with that, given they're the best representation of how to easily use the underlying library code?

**Reply by Bibo-Joshi** [reply]
>There's no code examples. It's all rather theoretical.

Well, as the wiki page states:

> There is no one way to handle networking issues. Each bot has its own usage pattern and unique network behaviour.

😉

> Shouldn't the examples deal with that, given they're the best representation of how to easily use the underlying library code?

The paragraphs on `urllib3` tweaks, `getUpdates` and socket opts all bascially just describe parameters that you can pass - IMO there is no need to provide more elaborate examples on how you pass parameters to functions in python.

The [last paragraph](https://github.com/python-telegram-bot/python-telegram-bot/wiki/Handling-network-errors#stabilizing-your-app) basically says that some `try-except` can be useful. `try-except` is again a general python concept so IMO it's not necessary to elaborate it here. What would probably help here is to link to the [exception classes](https://python-telegram-bot.readthedocs.io/en/stable/telegram.error.html) of PTB and I just edited the wiki page to do so :)

> This is what I tried, thus far. To no avail:

error handlers only get information about the exception, not about which method raised it. to retry calling a bot method, error handlers hence are of little use.

**Reply by Happy-Ferret** [reply]
Ok. So what is an efficient way to retry the same method then?
That's what I'm missing.

A while loop?

**Reply by Bibo-Joshi** [reply]
well, that's the part that depends on your use case. If you want to be absolutely sure that the method executed successfully, you can use a `while` loop and break when it worked. But there is no deterministic stopping criterion, which is a downside. you could also use a `for` loop to try at most n times. you may want to pause a bit between tries, you may want to change the timeout on the second try (maybe based on which error you got) and so on and so on.
Think about which behavior would be beneficial for you and implement it :)

**Reply by Happy-Ferret** [reply]
Ok. So if I understand this correctly, the error handler is the wrong place to implement a retry logic, correct?
I guess I could write a `@retry(retries: int, sleep: int)` decorator.

I was hoping for a more general solution. Something akin to Telebot's `infinity_polling`.

**Reply by Bibo-Joshi** [reply]
> Ok. So if I understand this correctly, the error handler is the wrong place to implement a retry logic, correct?

yes

> I guess I could write a `@retry(retries: int, sleep: int)` decorator.

Note that this would retry the whole callback, which may be too much. e.g. if your callback sends 2 messages and only the second one fails, then a `retry` decorator would call the whole callback again, leading to the first message also being send again even though it was successful the first time already

> I was hoping for a more general solution. Something akin to Telebot's `infinity_polling`.

You're mixing things up here: `infinty_polling` is a long running task to __fetch__ updates, not to handle them. `Updater.start_polling`. PTBs equivalent is `Updater.start_polling`, which is completely independent of `Dispatcher`, who manages the handlers

**Reply by Happy-Ferret** [reply]
> Note that this would retry the whole callback, which may be too much. e.g. if your callback sends 2 messages and only the second one fails, then a retry decorator would call the whole callback again, leading to the first message also being send again even though it was successful the first time already

Yep. You're right. I thought about that right after I posted it. Lol.

Hm. I guess unless I'm overlooking something obvious, the best solution would be to wrap the most recurring API calls in my own functions. I'd rather not clutter my entire codebase with try-except logic.

> You're mixing things up here: infinty_polling is a long running task to fetch updates, not to handle them. Updater.start_polling. PTBs equivalent is Updater.start_polling, which is completely independent of Dispatcher, who manages the handlers

Oh. Ok. I wasn't sure. I thought infinity_polling was basically a general solution that the creator of Telebot built out of PTB.

**Reply by Bibo-Joshi** [reply]
> creator of Telebot built out of PTB.

not sure if I'm understanding you correctly, but pyTelegramBotAPI is not built on top of PTB - it's an independent library

**Reply by Happy-Ferret** [reply]
> > creator of Telebot built out of PTB.
> 
> not sure if I'm understanding you correctly, but pyTelegramBotAPI is not built on top of PTB - it's an independent library

My bad. I could've sworn I saw references to PTB.
I guess I was browsing with too many tabs. Lol

**Comment by Bibo-Joshi** [comment]
Note that `time.sleep(10)` will make your bot unresponsive for 10 seconds. Maybe that's what you want - maybe it isn't. Just pointing it out.

**Reply by Happy-Ferret** [reply]
> Note that `time.sleep(10)` will make your bot unresponsive for 10 seconds. Maybe that's what you want - maybe it isn't. Just pointing it out.

That's what I figured. I just found it in another Telegram bot library that has an `infinity_polling` method. It didn't make sense to me at the time but I added it for trial-and-error.

**Comment by Happy-Ferret** [comment, marked as answer]
This is the solution I ended up settling on.

```python
def retry_on_error(func, wait=0.1, retry=0, *args, **kwargs):
    i = 0
    while True:
        try:
            return func(*args, **kwargs)
            break
        except telegram.error.NetworkError:
            logging.exception(f"Network Error. Retrying...{i}")
            i += 1
            time.sleep(wait)
            if retry != 0 and i == retry:
                break
```
Which can then be used as follows:

```python
retry_on_error(update.message.reply_photo, photo=url)
```

EDIT: Should be `return func...`

**Reply by Poolitzer** [reply]
time.sleep() stops the whole thread, which might be good and maybe not depending on your code, and I would have probably used a for loop, but very neat solution! Thanks for sharing.

**Reply by Vihagha-Dinethma** [reply]
@Poolitzer can you share your final code please
